File: repodiff/utils/token_counter.py
import tiktoken
from typing import Optional


class TokenCounter:
    """
    Handles token counting for LLM models using OpenAI's tiktoken.
    """
    
    def __init__(self, model: str = "gpt-4o"):
        """
        Initialize the TokenCounter with a specific model.
        
        Args:
            model: The name of the LLM model to use for token counting.
        """
        self.model = model
        self.encoding = tiktoken.encoding_for_model(model)
    
    def count_tokens(self, text: str) -> int:
        """
        Count the number of tokens in the given text.
        
        Args:
            text: The text to count tokens for.
            
        Returns:
            The number of tokens in the text.
        """
        return len(self.encoding.encode(text)) 

File: src/utils/token_counter.rs
use tiktoken_rs::CoreBPE;
use crate::error::{RepoDiffError, Result};

/// Handles token counting for LLM models using tiktoken
pub struct TokenCounter {
    /// The tiktoken encoding
    bpe: CoreBPE,
}

impl TokenCounter {
    /// Initialize the TokenCounter with a specific model
    ///
    /// # Arguments
    ///
    /// * `model` - The name of the LLM model to use for token counting
    pub fn new(model: &str) -> Result<Self> {
        let bpe = tiktoken_rs::get_bpe_from_model(model)
            .map_err(|e| RepoDiffError::TiktokenError(format!("Failed to get BPE for model {}: {}", model, e)))?;
        Ok(Self { bpe })
    }

    /// Count the number of tokens in the given text
    ///
    /// # Arguments
    ///
    /// * `text` - The text to count tokens for
    pub fn count_tokens(&self, text: &str) -> usize {
        self.bpe.encode_ordinary(text).len()
    }
} 